### PR TITLE
8294062: Improve parsing performance of j.l.c.MethodTypeDesc

### PR DESCRIPTION
**Pull request by asotona**
Parsing performance of `j.l.c.MethodTypeDescriptor::ofDescriptor` can be significantly improved and add performance boost to intensive users.

Two main reasons have been identified and fixed in this patch:

First glitch is late insertion of return type to the first position of `j.u.ArrayList`, causing all previously parsed parameter types to shift using `System::arraycopy`.

Proposed patch inserts a placeholder for the return type first and replaces it with return type later.

Second performance degradation is in `MethodTypeDescriptor::skipOverFieldSignature`, where each identified class descriptor is exactly located by `String::indexOf`, extracted by `String::substring`, and passed for for the purpose of verification to `ConstantUtils::verifyUnqualifiedClassName`.

Proposed patch inlines the unqualified class name verification into `ConstantUtils::skipOverFieldSignature` parsing loop and so eliminates calls of `String::indexOf` and `String::substring`. Package private method `ConstantUtils::verifyUnqualifiedClassName` has no other use within the package, so it is removed.

Please review proposed performance improving patch.

Thanks,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294062](https://bugs.openjdk.org/browse/JDK-8294062): Improve parsing performance of j.l.c.MethodTypeDesc

### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10358/head:pull/10358` \
`$ git checkout pull/10358`

Update a local copy of the PR: \
`$ git checkout pull/10358` \
`$ git pull https://git.openjdk.org/jdk pull/10358/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10358`

View PR using the GUI difftool: \
`$ git pr show -t 10358`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10358.diff">https://git.openjdk.org/jdk/pull/10358.diff</a>

</details>

### CODE REVIEW

**Comment by asotona** [comment]
JMH-measured performance boost of the patch when parsing single MethodTypeDesc with 8 Objects as method arguments is 33%.

File: src/java.base/share/classes/java/lang/constant/ConstantUtils.java [hunk 1]
```diff
@@ -203,16 +204,22 @@ static int skipOverFieldSignature(String descriptor, int start, int end, boolean
...
+                    while (++index < end) {
+                        switch (descriptor.charAt(index)) {
+                            case ';' -> {
+                                return legal ? index - start + 1 : 0;
+                            }
+                            case '.', '[' -> {
+                                return 0;
+                            }
+                            case '/' -> {
+                                if (!legal) return 0;
```
[the remark this answers is not in the document]
Comment (asotona) [reply src/java.base/share/classes/java/lang/constant/ConstantUtils.java:217]: `legal` is initially set to false, so leading `/` is also rejected

[the remark this answers is not in the document]
Comment (PaulSandoz) [reply src/java.base/share/classes/java/lang/constant/ConstantUtils.java:217]: Doh! Could you add a comment with the case block, something like?
// Rejects a leading `/`, any "//", and a trailing '/' (by exit from the loop)

[the remark this answers is not in the document]
Comment (asotona) [reply src/java.base/share/classes/java/lang/constant/ConstantUtils.java]: I've added the comments, thanks.

**Review by PaulSandoz — APPROVED** [review]
Looks good.

**Pull request closed without merging**